local opam switch create could use better compiler selection logic

With a local switch:

```
$ opam switch create .
[NOTE] Converting format of /Users/avsm/src/git/avsm/ocaml-dockerfile/dockerfile-cmd.opam from 1.2 to 2.0~beta5
[NOTE] Converting format of /Users/avsm/src/git/avsm/ocaml-dockerfile/dockerfile-opam.opam from 1.2 to 2.0~beta5
[NOTE] Converting format of /Users/avsm/src/git/avsm/ocaml-dockerfile/dockerfile.opam from 1.2 to 2.0~beta5
[WARNING] No possible installation was found including a compiler and the selected packages.
Create the switch with no specific compiler selected, and attempt to continue anyway ? [Y/n] 
```

Selecting yes will result in 4.07.0+variants+trunk or something like that being selected, rather than the latest stable compiler.

Then

```
$ opam switch create . 4.06.0
[ERROR] Compiler selection '4.06.0' is ambiguous. matching packages: { ocaml-base-compiler.4.06.0, ocaml-system.4.06.0 }
```

fails due to the ambiguity, and then

```
$ opam switch create . ocaml-system.4.06.0
```

finally works.  These all seem like common failure scenarios -- having multiple compilers installed, not specifying a default and so on.  Shouldn't the local switch create default to the system compiler instead, for example?

> `[WARNING] No possible installation was found including a compiler and the selected packages.`

Hmm, this sounds like a bug, I'll check. I doubt it's true :), and if it's not, opam should automatically choose that compiler.

> ```
> $ opam switch create . 4.06.0
> [ERROR] Compiler selection '4.06.0' is ambiguous. matching packages: { ocaml-base-compiler.4.06.0, ocaml-system.4.06.0 }
> ```

This one I don't see how to avoid. Remember that specifying a version with `opam switch create` is a semi-hack for backwards compat, what opam needs is a package name(.opt version) that has the `compiler` flag set.

Thanks for reporting, the above bug seems annoying indeed!

--- a/release/Dockerfile.in
+++ b/release/Dockerfile.in
@@ -3,7 +3,7 @@ FROM multiarch/debian-debootstrap:%TARGET_TAG%
 #    docker run --rm --privileged multiarch/qemu-user-static:register --reset
 LABEL Description="opam release builds" Vendor="OCamlPro" Version="1.0"
 
-RUN apt-get update && apt-get install bzip2 g++ make patch wget libglpk-dev libltdl-dev --yes && apt-get clean --yes
+RUN apt-get update && apt-get install bzip2 g++ make patch wget libltdl-dev --yes && apt-get clean --yes
 RUN useradd -U --create-home opam
 
 ADD https://caml.inria.fr/pub/distrib/ocaml-4.04/ocaml-4.04.2.tar.gz /root/

--- a/release/Makefile
+++ b/release/Makefile
@@ -45,13 +45,13 @@ build/%.image: build/Dockerfile.%
 # Actually, this is for debian 8 jessie, and varies wildly
 CLINKING_linux = \
 -Wl,-Bstatic \
--lunix -lbigarray -lmccs_stubs \
+-lunix -lbigarray -lmccs_stubs -lmccs_glpk_stubs \
 -lstdc++ \
 -Wl,-Bdynamic \
 -static-libgcc
 
 CLINKING_darwin = \
--lunix -lbigarray -lmccs_stubs \
+-lunix -lbigarray -lmccs_stubs -lmccs_glpk_stubs \
 -lstdc++
 
 CLINKING_openbsd = $(CLINKING_darwin)

--- a/release/release.sh
+++ b/release/release.sh
@@ -2,7 +2,7 @@
 
 # This script is expected to run on Linux with docker available, and to have two
 # remotes "some-osx" and "some-openbsd", with the corresponding OSes, ocaml deps
-# and glpk installed
+# installed
 
 DIR=$(dirname $0)
 cd "$DIR"

--- a/src/client/opamAuxCommands.ml
+++ b/src/client/opamAuxCommands.ml
@@ -441,7 +441,7 @@ let get_compatible_compiler ?repos ?locked rt dir =
       packages =
          virt_st.packages ++ local_packages ++ pin_depends;
       available_packages =
-        lazy (Lazy.force available ++ pin_depends);
+        lazy (Lazy.force available ++ local_packages ++ pin_depends);
     }
   in
   let univ =
@@ -461,11 +461,21 @@ let get_compatible_compiler ?repos ?locked rt dir =
           (OpamSwitchState.opam virt_st nv))
       (OpamPackage.Set.of_list alldeps)
   in
-  let compilers =
+  let installable =
     OpamSolver.installable_subset
       {univ with u_base = local_packages; u_installed = local_packages}
-      compilers
+      (OpamPackage.Set.union local_packages compilers)
   in
+  if OpamPackage.Set.is_empty installable then
+    (OpamConsole.error
+       "The following local packages don't appear to be installable:\n%s"
+       (OpamStd.Format.itemize OpamPackage.to_string
+          (OpamPackage.Set.elements local_packages));
+     if OpamConsole.confirm "Do you want to create an empty switch nonetheless ?"
+     then []
+     else OpamStd.Sys.exit_because `Aborted)
+  else
+  let compilers = OpamPackage.Set.inter compilers installable in
   try
     [OpamSolution.eq_atom_of_package
        (OpamPackage.Set.choose_one compilers)]

--- a/src/client/opamAuxCommands.ml
+++ b/src/client/opamAuxCommands.ml
@@ -276,7 +276,7 @@ let simulate_local_pinnings st to_pin =
           let opam, version = match OpamFile.OPAM.version_opt opam with
             | Some v -> opam, v
             | None ->
-              let v = OpamPackage.Version.of_string "~dev" in
+              let v = OpamPinCommand.default_version st name in
               OpamFile.OPAM.with_version v opam, v
           in
           OpamPackage.Map.add (OpamPackage.create name version) opam map)
@@ -388,6 +388,9 @@ let autopin st ?(simulate=false) ?quiet ?locked atom_or_local_list =
 
 let get_compatible_compiler ?repos ?locked rt dir =
   let gt = rt.repos_global in
+  let virt_st =
+    OpamSwitchState.load_virtual ?repos_list:repos gt rt
+  in
   let local_files = opams_of_dir ?locked dir in
   let local_opams =
     List.fold_left (fun acc (name, f) ->
@@ -397,7 +400,7 @@ let get_compatible_compiler ?repos ?locked rt dir =
           match OpamFile.OPAM.version_opt opam with
           | Some v -> OpamPackage.create name v, opam
           | None ->
-            let v = OpamPackage.Version.of_string "~dev" in
+            let v = OpamPinCommand.default_version virt_st name in
             OpamPackage.create name v,
             OpamFile.OPAM.with_version v opam
         in
@@ -416,9 +419,6 @@ let get_compatible_compiler ?repos ?locked rt dir =
     OpamSolution.eq_atoms_of_packages local_packages
   in
   let virt_st =
-    let virt_st =
-      OpamSwitchState.load_virtual ?repos_list:repos gt rt
-    in
     let opams =
       OpamPackage.Map.union (fun _ x -> x) virt_st.opams local_opams
     in

--- a/src/client/opamPinCommand.ml
+++ b/src/client/opamPinCommand.ml
@@ -337,6 +337,10 @@ exception Nothing_to_do
 end
 include Exns
 
+let default_version st name =
+  try OpamPackage.version (OpamSwitchState.get_package st name)
+  with Not_found -> OpamPackage.Version.of_string "~dev"
+
 let rec handle_pin_depends st nv opam =
   let extra_pins = OpamFile.OPAM.pin_depends opam in
   let extra_pins =
@@ -373,12 +377,6 @@ and source_pin
     (slog OpamPackage.Name.to_string) name
     (slog (OpamStd.Option.to_string OpamPackage.Version.to_string)) version
     (slog (OpamStd.Option.to_string ~none:"none" OpamUrl.to_string)) target_url;
-  let installed_version =
-    try
-      Some (OpamPackage.version
-              (OpamSwitchState.find_installed_package_by_name st name))
-    with Not_found -> None
-  in
 
   let cur_version, cur_urlf =
     try
@@ -413,12 +411,7 @@ and source_pin
             "Are you sure you want to override this and pin it anyway ?"
         then raise Exns.Aborted
       );
-      let version =
-        try OpamPackage.version (OpamSwitchState.get_package st name)
-        with Not_found ->
-          OpamStd.Option.Op.(installed_version +!
-                             OpamPackage.Version.of_string "~dev")
-      in
+      let version = default_version st name in
       version, None
   in
 

--- a/src/client/opamPinCommand.mli
+++ b/src/client/opamPinCommand.mli
@@ -63,3 +63,7 @@ val list: 'a switch_state -> short:bool -> unit
     it *)
 val read_opam_file_for_pinning:
   name -> OpamFile.OPAM.t OpamFile.t -> url -> OpamFile.OPAM.t option
+
+(** The default version for pinning a package: depends on the state, what is
+    installed and available, and defaults to [~dev]. *)
+val default_version: 'a switch_state -> name -> version

--- a/src/state/opamFormatUpgrade.ml
+++ b/src/state/opamFormatUpgrade.ml
@@ -15,6 +15,7 @@ open OpamStd.Op
 open OpamFilename.Op
 
 let log fmt = OpamConsole.log "FMT_UPG" fmt
+let slog = OpamConsole.slog
 
 (* - Package and aux functions - *)
 
@@ -1079,9 +1080,10 @@ let opam_file ?(quiet=false) ?filename opam =
   then
     ((match filename with
         | Some f when not quiet ->
-          OpamConsole.note "Converting format of %s from %s to %s"
-            (OpamFile.to_string f) (OpamVersion.to_string v)
-            (OpamVersion.to_string latest_version)
+          log "Internally converting format of %a from %a to %a"
+            (slog OpamFile.to_string) f
+            (slog OpamVersion.to_string) v
+            (slog OpamVersion.to_string) latest_version
         | _ -> ());
      opam_file_from_1_2_to_2_0 ?filename opam)
   else opam

